### PR TITLE
feat(conditional): supports promises that are returned in the schema function

### PR DESCRIPTION
**Pull request by bas080**
That allows for other to perform an async request and calculate the schema when
it is done.

Also added a more fitting name for conditional named computed

### CODE REVIEW

File: src/component/framework/conditional.js [hunk 1]
```diff
@@ -27,3 +45,7 @@ export class Conditional {
   }
 
 }
+
+function isPromise(value) {
+  return (typeof value === 'object' && value.then);
```
Comment (doktordirk) [inline src/component/framework/conditional.js]: instanceof Promise?

**Comment by doktordirk** [comment]
can that be documented somewhere. else looks fine (besides comment)

File: src/component/framework/conditional.js (hunk 1 above)
Comment (bas080) [reply src/component/framework/conditional.js]: Feels like that would require people to use Promise constructor of the global. I would want people to decide for themselves what promise lib they would like to use.

**Comment by bas080** [comment]
I'll document it shortly

File: src/component/framework/conditional.js (hunk 1 above)
Comment (doktordirk) [reply src/component/framework/conditional.js]: we do write esNext (or maybe typescript in future), and that has Promises, They are all over the place and if needed already added by polyfills to global. so, i think that concern is valid.

File: doc/types.md [hunk 2]
```diff
@@ -377,3 +380,34 @@ in the model or other values also.
...
+    type  : 'computed',
+    schema: () => {
+      return Promise((resolve, reject) => {
+        http.get('languages').then(languages => {
+          let schema = [
+            type: 'options',
+            options: languages.map(language => language.name);
+          ];
+
+          resolve(sche,a);
```
Comment (doktordirk) [inline doc/types.md]: typo

File: doc/types.md [hunk 3]
```diff
@@ -377,3 +380,34 @@ in the model or other values also.
...
+  };
+
+  let userSchema = [{
+    key: 'name',
+  }, {
+    key   : 'language',
+    type  : 'computed',
+    schema: () => {
+      return Promise((resolve, reject) => {
+        http.get('languages').then(languages => {
```
Comment (doktordirk) [inline doc/types.md]: in this example you can just return the  http.get('languages').then(l... directly. no need to another promise or catching

schema: () => http.get('languages').then ...

File: src/component/framework/conditional.js [hunk 4]
```diff
@@ -20,8 +20,6 @@ export class Conditional {
         return schema;
       }
 
-      // support promise
-      // @todo check for type
       if (isPromise(schema)) {
         return schema.then(resolved => {
           this.schema = resolved;
```
Comment (doktordirk) [inline src/component/framework/conditional.js:25]: 1 line beloow: here you should catch and log

File: src/component/framework/conditional.js [hunk 5]
```diff
@@ -47,5 +45,5 @@ export class Conditional {
 }
 
 function isPromise(value) {
```
Comment (doktordirk) [inline src/component/framework/conditional.js:47]: not sure if that is still worth it. i'd inline it

File: doc/types.md [hunk 6]
```diff
@@ -396,16 +396,11 @@ The schema function also allows the returning of Promises. e.g.
...
-            options: languages.map(language => language.name);
-          ];
-
-          resolve(sche,a);
-        })
-        .catch(reject);
+      return http.get('languages').then(languages => {
+        return [
+          type: 'options',
+          options: languages.map(language => language.name);
```
Comment (doktordirk) [inline doc/types.md]: typo, remove semi-colon

File: doc/types.md [hunk 7]
```diff
@@ -396,16 +396,11 @@ The schema function also allows the returning of Promises. e.g.
...
+        ];
       });
```
Comment (doktordirk) [inline doc/types.md:404]: it's fine to leave it like this for maybe increased readability. but i want to point out that one also can write shorter

``` js
schema: () => http.get('languages').then(languages => [
  type: 'options',
  options: languages.map(language => language.name)
   ]);
```

**Comment by jeremyvergnas** [comment]
Looks good to me 👍

**Pull request merged**